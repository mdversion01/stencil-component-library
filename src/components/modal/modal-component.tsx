// src/components/modal/modal-component.tsx
import { Component, h, Prop, Element, Method } from '@stencil/core';
// ✅ Import the Modal class directly from Bootstrap ESM
import Modal from 'bootstrap/js/dist/modal';

@Component({
  tag: 'modal-component',
  styleUrls: ['modal.scss', '../button/button.scss'],
  shadow: false,
})
export class ModalComponent {
  @Element() host!: HTMLElement;

  // Button Props
  @Prop() ariaLabel: string = '';
  @Prop() block: boolean = false;
  @Prop() btnText = 'Launch demo modal';
  @Prop() cancelCloseBtn: string = 'Close';
  @Prop() classNames: string = '';
  @Prop() disabled: boolean = false;
  @Prop() link: boolean = false;
  @Prop() outlined: boolean = false;
  @Prop() ripple: boolean = false;
  @Prop() size: string = '';
  @Prop() shape: string = '';
  @Prop() titleAttr: string = '';
  @Prop() variant = '';

  // Modal Props
  @Prop() modalId?: string;
  @Prop() modalTitle = 'Modal title';
  @Prop({ reflect: true }) modalSize?: 'sm' | 'lg' | 'xl';
  @Prop({ reflect: true }) modalFullScreen?: 'fullscreen' | 'sm-down' | 'md-down' | 'lg-down' | 'xl-down' | 'xxl-down';
  @Prop() scrollableBody: boolean = false;
  @Prop() scrollLongContent: boolean = false;
  @Prop() verticallyCentered: boolean = false;

  private modalEl?: HTMLDivElement;
  private modalInstance?: Modal;
  private lastTrigger?: HTMLElement;

  private get id() {
    return this.modalId || `modal-component-${Math.random().toString(36).slice(2, 9)}`;
  }

  // ✅ Instantiate once, after the element exists in the DOM
  componentDidLoad() {
    if (!this.modalEl) return;

    this.modalInstance = new Modal(this.modalEl, { backdrop: true, keyboard: true, focus: true });

    // During hide: blur anything focused inside and temporarily set inert
    this.modalEl.addEventListener('hide.bs.modal', () => {
      const ae = document.activeElement as HTMLElement | null;
      if (ae && this.modalEl!.contains(ae)) ae.blur();
      this.modalEl!.setAttribute('inert', '');
    });

    // After fully hidden: remove inert and return focus to the trigger
    this.modalEl.addEventListener('hidden.bs.modal', () => {
      this.modalEl!.removeAttribute('inert');
      this.lastTrigger?.focus?.();
    });

    // After fully shown: ensure dialog/focus is safe
    this.modalEl.addEventListener('shown.bs.modal', () => {
      (this.modalEl!.querySelector('.modal-dialog') as HTMLElement)?.focus?.();
    });
  }

  disconnectedCallback() {
    // ✅ Clean up when removed
    try {
      (this.modalInstance as any)?.dispose?.();
    } catch {}
    this.modalInstance = undefined;
  }

  /** Open programmatically */
  @Method() async open() {
    this.modalInstance?.show();
  }

  /** Close programmatically */
  @Method() async close() {
    // Important: blur focus inside the modal before hide
    const ae = document.activeElement as HTMLElement | null;
    if (ae && this.modalEl?.contains(ae)) ae.blur();
    this.modalInstance?.hide();
  }

  private onTriggerClick = (e: MouseEvent) => {
    this.lastTrigger = e.currentTarget as HTMLElement;
    this.open();
  };

  render() {
    const id = this.id;

    const classList = [
      'btn',
      this.block && 'btn--block',
      this.classNames,
      this.link && 'link',
      this.outlined ? `btn-outline-${this.variant}` : '',
      !this.outlined && this.variant ? `btn-${this.variant}` : '',

      this.size && `btn-${this.size}`,
      this.shape && `btn-${this.shape}`,
      this.ripple && 'btn-ripple',
    ]
      .filter(Boolean)
      .join(' ');

    const computedAriaLabel =
      (this.ariaLabel && this.ariaLabel.trim()) ||
      // (hasVisibleText ? this.btnText.trim() : '') ||
      (this.titleAttr && this.titleAttr.trim()) ||
      'Button';

    const buttonProps = {
      'aria-label': computedAriaLabel,
      'aria-disabled': this.link && this.disabled ? 'true' : undefined,
      'class': classList,
      'disabled': !this.link && this.disabled,
    };

    return (
      <div class="modal-component-wrapper">
        {/* Trigger button */}
        <button type="button" role="button" data-bs-toggle="modal" data-bs-target={`#${id}`} {...(buttonProps as any)} onClick={this.onTriggerClick}>
          {this.btnText}
        </button>

        {/* Modal (Bootstrap 5 markup) */}
        <div
          class={`modal fade ${this.scrollLongContent ? ' overflow-x: hidden; overflow-y: auto;' : ''}`}
          id={id}
          tabIndex={-1}
          aria-labelledby={`${id}-label`}
          aria-hidden="true"
          role="dialog"
          ref={el => (this.modalEl = el as HTMLDivElement)}
        >
          <div
            class={`modal-dialog ${this.modalSize ? ` modal-${this.modalSize}` : ''}${this.modalFullScreen ? ` modal-${this.modalFullScreen}` : ''}${this.verticallyCentered ? ' modal-dialog-centered' : ''}`}
            tabIndex={-1}
            role="document"
          >
            <div class="modal-content">
              <div class="modal-header">
                <h5 class="modal-title" id={`${id}-label`}>
                  {this.modalTitle}
                </h5>
                <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close" onClick={() => this.close()} />
              </div>

              <div class={` modal-body ${this.scrollableBody ? ' scrollable' : ''}`}>
                <slot></slot>
              </div>

              <div class="modal-footer">
                <button type="button" class={`btn btn-secondary ${this.size ? `btn-${this.size}` : ''}`} data-bs-dismiss="modal" onClick={() => this.close()}>
                  {this.cancelCloseBtn}
                </button>
                <slot name="footer"></slot>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}
