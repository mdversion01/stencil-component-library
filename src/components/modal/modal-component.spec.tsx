// src/components/modal/modal-component.spec.tsx
import { newSpecPage } from '@stencil/core/testing';
import { ModalComponent } from './modal-component';

// --- Helpers: stabilize random ids & normalize HTML for snapshots ---
const ID_REGEX = /modal-component-[a-z0-9]{7}/g;
function normalizeHtml(node: Element | ShadowRoot | DocumentFragment | string) {
  const html = typeof node === 'string' ? node : (node as any).outerHTML || (node as any).innerHTML || '';
  return html.replace(ID_REGEX, 'modal-component-<id>');
}

let randSpy: jest.SpyInstance<number, []>;

// --- Mock Bootstrap Modal (default export) ---
const showMock = jest.fn();
const hideMock = jest.fn();
const disposeMock = jest.fn();

jest.mock('bootstrap/js/dist/modal', () => {
  const MockCtor = function (this: any) {
    // no-op; instance methods below
  } as any;

  MockCtor.prototype.show = showMock;
  MockCtor.prototype.hide = hideMock;
  MockCtor.prototype.dispose = disposeMock;

  return {
    __esModule: true,
    default: MockCtor,
  };
});

describe('modal-component', () => {
  beforeEach(() => {
    showMock.mockClear();
    hideMock.mockClear();
    disposeMock.mockClear();
    randSpy = jest.spyOn(Math, 'random').mockReturnValue(0.9999999); // => â€¦zzzzzzz
  });

  afterEach(() => {
    randSpy?.mockRestore();
  });

  it('renders trigger + modal markup (snapshot)', async () => {
    const page = await newSpecPage({
      components: [ModalComponent],
      html: `<modal-component btn-text="Launch demo modal" variant="primary">
               <p>Body content</p>
               <div slot="footer"><button class="btn btn-primary">Save</button></div>
             </modal-component>`,
    });

    // Snapshot the rendered host
    expect(normalizeHtml(page.root as any)).toMatchSnapshot('render-default');
  });

  it('applies dialog size, fullscreen and centered classes (snapshot)', async () => {
    const page = await newSpecPage({
      components: [ModalComponent],
      html: `<modal-component
               btn-text="Open"
               modal-title="Sized"
               modal-size="lg"
               modal-full-screen="md-down"
               vertically-centered="true">
             </modal-component>`,
    });

    const dialog = page.root!.querySelector('.modal-dialog') as HTMLElement;
    expect(dialog).toBeTruthy();
    // Should include: modal-lg, modal-md-down, modal-dialog-centered
    expect(dialog.className).toEqual(expect.stringContaining('modal-lg'));
    expect(dialog.className).toEqual(expect.stringContaining('modal-md-down'));
    expect(dialog.className).toEqual(expect.stringContaining('modal-dialog-centered'));

    // Snapshot the modal subtree only (keeps snapshot tidy)
    const modal = page.root!.querySelector('.modal') as HTMLElement;
    expect(normalizeHtml(modal)).toMatchSnapshot('render-sized-centered');
  });

  it('opens programmatically (calls Bootstrap show)', async () => {
    const page = await newSpecPage({
      components: [ModalComponent],
      html: `<modal-component modal-title="Programmatic"></modal-component>`,
    });

    const instance = page.rootInstance as ModalComponent;

    // Make the call deterministic in JSDOM by stubbing the Bootstrap instance
    showMock.mockClear?.();
    hideMock.mockClear?.();
    disposeMock.mockClear?.();
    (instance as any).modalInstance = {
      show: showMock,
      hide: hideMock,
      dispose: disposeMock,
    };

    await instance.open();

    expect(showMock).toHaveBeenCalledTimes(1);
  });

  it('closes programmatically (calls Bootstrap hide)', async () => {
    const page = await newSpecPage({
      components: [ModalComponent],
      html: `<modal-component modal-title="Programmatic"></modal-component>`,
    });

    const instance = page.rootInstance as ModalComponent;

    // Ensure the component is fully rendered
    await page.waitForChanges();

    // ðŸ”§ Inject a mock Bootstrap Modal instance so open()/close() are deterministic
    // If you already have showMock/hideMock/disposeMock at spec scope, this will reuse them.
    showMock.mockClear?.();
    hideMock.mockClear?.();
    disposeMock.mockClear?.();

    (instance as any).modalInstance = {
      show: showMock,
      hide: hideMock,
      dispose: disposeMock,
    };

    await instance.open();
    await instance.close();

    expect(showMock).toHaveBeenCalledTimes(1);
    expect(hideMock).toHaveBeenCalledTimes(1);
  });

  it('close button inside footer calls close()', async () => {
    const page = await newSpecPage({
      components: [ModalComponent],
      html: `<modal-component modal-title="Close Test">
             <div slot="footer">
               <button class="btn btn-secondary" id="footer-close">Close</button>
             </div>
           </modal-component>`,
    });

    const instance = page.rootInstance as any;

    // Ensure component lifecycle/refs have settled
    await page.waitForChanges();

    // ðŸ”§ Inject a deterministic mock instance so .open()/.close() work even if
    // Bootstrap's constructor didn't run in this test environment.
    instance.modalInstance = {
      show: showMock,
      hide: hideMock,
      dispose: disposeMock,
    };

    await (instance as ModalComponent).open();
    expect(showMock).toHaveBeenCalledTimes(1);

    // (Optional) pretend user wired a click handler on their footer button to call el.close().
    // We directly call close() here to validate the component API.
    await (instance as ModalComponent).close();
    expect(hideMock).toHaveBeenCalledTimes(1);
  });
});
