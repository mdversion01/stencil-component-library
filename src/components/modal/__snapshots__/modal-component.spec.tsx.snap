// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modal-component applies dialog size, fullscreen and centered classes (snapshot): render-sized-centered 1`] = `"<div class="modal fade" id="modal-component-<id>" tabindex="-1" aria-labelledby="modal-component-<id>-label" aria-hidden="true" role="dialog"><div class="modal-dialog modal-lg modal-md-down modal-dialog-centered" tabindex="-1" role="document"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="modal-component-<id>-label">Sized</h5><button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button></div><div class="modal-body">  </div><div class="modal-footer"><button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Close</button> </div></div></div></div>"`;

exports[`modal-component renders trigger + modal markup (snapshot): render-default 1`] = `"<modal-component btn-text="Launch demo modal" variant="primary"><!---->     <div class="modal-component-<id>"><button type="button" role="button" data-bs-toggle="modal" data-bs-target="#modal-component-<id>" aria-label="Button" class="btn btn-primary">Launch demo modal</button><div class="modal fade" id="modal-component-<id>" tabindex="-1" aria-labelledby="modal-component-<id>-label" aria-hidden="true" role="dialog"><div class="modal-dialog" tabindex="-1" role="document"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="modal-component-<id>-label">Modal title</h5><button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button></div><div class="modal-body">  <p>Body content</p>  </div><div class="modal-footer"><button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Close</button> <div slot="footer"><button class="btn btn-primary">Save</button></div></div></div></div></div></div></modal-component>"`;
